solr/core/src/java/org/apache/solr/search/KeepAliveRegenerator.java

```diff
@@ -0,0 +1,593 @@
+          }
+
+          @Override
+          public boolean acceptField(String field) {
+            return !ret[0];
+          }
+        });
+    return ret[0];
+  }
+
+  private static class KeepAliveSegAwareValue
+      implements MetaEntry<Query, DocSet, KeepAliveSegAwareValue> {
+
+    private static final long BASE_RAM_BYTES_USED =
+        RamUsageEstimator.shallowSizeOfInstance(KeepAliveSegAwareValue.class)
+            + RamUsageEstimator.shallowSizeOfInstance(AtomicReference.class)
+            + RamUsageEstimator.shallowSizeOfInstance(AbstractMap.SimpleImmutableEntry.class)
+            + RamUsageEstimator.shallowSizeOfInstance(CompletableFuture.class);
+
+    private final AtomicReference<
+            AbstractMap.SimpleImmutableEntry<SegmentMap, CompletableFuture<DocSet>>>
+        ref;
+    private final LongAdder partialHits;
+    private final DoubleAdder partialHitsRatio;
+    private final long ramBytesUsed;
+    private long accessTimestampNanos;
+
+    public KeepAliveSegAwareValue(
+        SegmentMap segMap, DocSet val, LongAdder partialHits, DoubleAdder partialHitsRatio) {
+      CompletableFuture<DocSet> f = new CompletableFuture<>();
+      f.complete(val);
+      ref = new AtomicReference<>(new AbstractMap.SimpleImmutableEntry<>(segMap, f));
+      ramBytesUsed = BASE_RAM_BYTES_USED + val.ramBytesUsed();
+      this.accessTimestampNanos = System.nanoTime();
+      this.partialHits = partialHits;
+      this.partialHitsRatio = partialHitsRatio;
+    }
+
+    public KeepAliveSegAwareValue(
+        SegmentMap segMap,
+        DocSet val,
+        long accessTimestampNanos,
+        LongAdder partialHits,
+        DoubleAdder partialHitsRatio) {
+      CompletableFuture<DocSet> f = new CompletableFuture<>();
+      f.complete(val);
+      ref = new AtomicReference<>(new AbstractMap.SimpleImmutableEntry<>(segMap, f));
+      ramBytesUsed = BASE_RAM_BYTES_USED + val.ramBytesUsed();
+      this.accessTimestampNanos = accessTimestampNanos;
+      this.partialHits = partialHits;
+      this.partialHitsRatio = partialHitsRatio;
+    }
+
+    public KeepAliveSegAwareValue(
+        KeepAliveSegAwareValue template, LongAdder partialHits, DoubleAdder partialHitsRatio) {
+      AbstractMap.SimpleImmutableEntry<SegmentMap, CompletableFuture<DocSet>> entry =
+          template.ref.get();
+      ref = new AtomicReference<>(template.ref.get());
+      DocSet val = entry.getValue().getNow(null);
+      if (val == null) {
+        // warming a value that's not yet loaded; just use the old `ramBytesUsed`
+        ramBytesUsed = template.ramBytesUsed();
+      } else {
+        ramBytesUsed = BASE_RAM_BYTES_USED + val.ramBytesUsed();
+      }
+      this.accessTimestampNanos = template.accessTimestampNanos;
+      this.partialHits = partialHits;
+      this.partialHitsRatio = partialHitsRatio;
+    }
+
+    @Override
+    public KeepAliveSegAwareValue metaClone(DocSet val) {
+      throw new UnsupportedOperationException();
```

Comment:
Perhaps a quick explanation why is this not supported? I suppose it's because `regenerateItem` is overridden in this class hence `metaClone` should NOT be called at all?

As from the class design, is there something preventing the support of this (other than not worth the effort if it's not used)